### DISCUSSION TITLE
Critic grad in the JAX implementation of the SAC algorithm

### DISCUSSION THREAD

**Discussion by MorningFrog** [Toni-SM/skrl#424] [General]
In the JAX implementation of the SAC algorithm (`skrl/agents/jax/sac/sac.py`), the `_update_critic` function returns only a single `grad`, which is then used to update both critic networks.

Source code:

```python
# https://jax.readthedocs.io/en/latest/faq.html#strategy-1-jit-compiled-helper-function
@functools.partial(jax.jit, static_argnames=("critic_1_act", "critic_2_act"))
def _update_critic(
    critic_1_act,
    critic_1_state_dict,
    critic_2_act,
    critic_2_state_dict,
    target_q1_values: jax.Array,
    target_q2_values: jax.Array,
    entropy_coefficient,
    next_log_prob,
    inputs: dict[str, jax.Array],
    sampled_rewards: jax.Array,
    sampled_terminated: jax.Array,
    discount_factor: float,
):
    # compute target values
    target_q_values = jnp.minimum(target_q1_values, target_q2_values) - entropy_coefficient * next_log_prob
    target_values = sampled_rewards + discount_factor * jnp.logical_not(sampled_terminated) * target_q_values

    # compute critic loss
    def _critic_loss(params, critic_act, role):
        critic_values, _ = critic_act(inputs, role=role, params=params)
        critic_loss = ((critic_values - target_values) ** 2).mean()
        return critic_loss, critic_values

    (critic_1_loss, critic_1_values), grad = jax.value_and_grad(_critic_loss, has_aux=True)(
        critic_1_state_dict.params, critic_1_act, "critic_1"
    )
    (critic_2_loss, critic_2_values), grad = jax.value_and_grad(_critic_loss, has_aux=True)(
        critic_2_state_dict.params, critic_2_act, "critic_2"
    )

    return grad, (critic_1_loss + critic_2_loss) / 2, critic_1_values, critic_2_values, target_values
```

And:

```python
            self.critic_1_optimizer = self.critic_1_optimizer.step(
                grad=grad, model=self.critic_1, lr=self.critic_learning_rate if self.critic_scheduler else None
            )
            self.critic_2_optimizer = self.critic_2_optimizer.step(
                grad=grad, model=self.critic_2, lr=self.critic_learning_rate if self.critic_scheduler else None
            )
```

This means both critics are updated using the gradient from `critic_2`, which seems counterintuitive.

Is this intentional, or is it an oversight?

**Comment by Toni-SM** [comment]
Hi @MorningFrog 

Yes, it's an issue I've had pending on skrl's project for some time :grimacing:  https://github.com/users/Toni-SM/projects/2/views/8?pane=issue&itemId=70840949

Thanks for bringing it to the forefront.
I am going to create an explicit bug for it

**Comment by Toni-SM** [comment]
Following it in https://github.com/Toni-SM/skrl/issues/425
